runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java

```diff
@@ -222,7 +282,117 @@ boolean canTranslate(T transform, FlinkStreamingTranslationContext context) {
+   * what was suggested on Flink's <a
+   * href="http://mail-archives.apache.org/mod_mbox/flink-dev/201901.mbox/%3CCAOUjMkygmFMDbOJNCmndrZ0bYug=iQJmVz6QvMW7C87n=pw8SQ@mail.gmail.com%3E">mailing
+   * list</a> and properly chooses shard keys in a way that they are distributed evenly among
+   * workers by Flink.
+   */
+  private static class FlinkAutoBalancedShardKeyShardingFunction<UserT, DestinationT>
+      implements ShardingFunction<UserT, DestinationT> {
+
+    private final int parallelism;
+    private final int maxParallelism;
+    private final Coder<DestinationT> destinationCoder;
+    private final Map<CacheKey, ShardedKey<Integer>> cache = new HashMap<>();
+    private final Set<Integer> usedSalts = new HashSet<>();
+
+    private int shardNumber = -1;
+
+    private FlinkAutoBalancedShardKeyShardingFunction(
+        int parallelism, int maxParallelism, Coder<DestinationT> destinationCoder) {
+      this.parallelism = parallelism;
+      this.maxParallelism = maxParallelism;
+      this.destinationCoder = destinationCoder;
+    }
+
+    @Override
+    public ShardedKey<Integer> assignShardKey(
+        DestinationT destination, UserT element, int shardCount) throws Exception {
+
+      if (shardNumber == -1) {
+        shardNumber = ThreadLocalRandom.current().nextInt(shardCount);
+      } else {
+        shardNumber = (shardNumber + 1) % shardCount;
+      }
+
+      int destinationKey =
+          Arrays.hashCode(CoderUtils.encodeToByteArray(destinationCoder, destination));
+      // we need to ensure that keys are always stable no matter at which worker they
+      // are created and what is an order of observed shard numbers
+      if (cache.size() < shardNumber) {
+        for (int i = 0; i < shardNumber; i++) {
+          generateInternal(new CacheKey(destinationKey, i));
+        }
+      }
+
+      return generateInternal(new CacheKey(destinationKey, shardNumber));
+    }
+
+    private ShardedKey<Integer> generateInternal(CacheKey key) {
+
+      ShardedKey<Integer> result = cache.get(key);
+      if (result != null) {
+        return result;
+      }
+
+      int salt = -1;
+      while (true) {
+        salt++;
+        ShardedKey<Integer> shk = ShardedKey.of(Objects.hash(key.key, salt), key.shard);
+        int targetPartition = key.shard % parallelism;
+
+        // create effective key in the same way Beam/Flink will do so we can see if it gets
+        // allocated to the partition we want
+        ByteBuffer effectiveKey;
+        try {
+          byte[] bytes = CoderUtils.encodeToByteArray(ShardedKeyCoder.of(VarIntCoder.of()), shk);
+          effectiveKey = ByteBuffer.wrap(bytes);
+        } catch (CoderException e) {
+          throw new RuntimeException(e);
+        }
+
+        int partition =
+            KeyGroupRangeAssignment.assignKeyToParallelOperator(
+                effectiveKey, maxParallelism, parallelism);
+
+        if (partition == targetPartition && !usedSalts.contains(salt)) {
```

Comment:
I know this will succeed by likelihood but I'm wondering whether we should put an upper limit here in case the loop does not return.